File: ActorMessagingExample/src/main/java/org/akka/essentials/hotswap/CalculatorActor.java
package org.akka.essentials.hotswap;

import akka.actor.UntypedActor;
import akka.japi.Procedure;

public class CalculatorActor extends UntypedActor {

	static final String ADD = "ADD";
	static final String SUBTRACT = "SUBTRACT";
	static final String MULTIPLY = "MULTIPLY";
	static final String DIVIDE = "DIVIDE";

	
	@Override
	public void preStart() {
		getContext().become(add);
	}

	@Override
	public void onReceive(Object message) throws Exception {

	}

	Procedure<Object> add = new Procedure<Object>() {
		public void apply(Object message) {
			if (message instanceof Numbers) {
				Numbers number = (Numbers) message;
				getSender().tell(number.first + number.second);
			} else if (message instanceof String) {
				checkAndUpdate((String) message);
			}
		}
	};

	Procedure<Object> subtract = new Procedure<Object>() {
		public void apply(Object message) {
			if (message instanceof Numbers) {
				Numbers number = (Numbers) message;
				getSender().tell(number.first - number.second);
			} else if (message instanceof String) {
				checkAndUpdate((String) message);
			}
		}
	};
	
	Procedure<Object> multiply = new Procedure<Object>() {
		public void apply(Object message) {
			if (message instanceof Numbers) {
				Numbers number = (Numbers) message;
				getSender().tell(number.first * number.second);
			} else if (message instanceof String) {
				checkAndUpdate((String) message);
			}
		}
	};
	
	Procedure<Object> divide = new Procedure<Object>() {
		public void apply(Object message) {
			if (message instanceof Numbers) {
				Numbers number = (Numbers) message;
				getSender().tell(number.first / number.second);
			} else if (message instanceof String) {
				checkAndUpdate((String) message);
			}
		}
	};

	private void checkAndUpdate(String msg) {
		if (msg.equals(ADD)) {
			getContext().unbecome();
			getContext().become(add);
		} else if (msg.equals(SUBTRACT)) {
			getContext().unbecome();
			getContext().become(subtract);
		} else if (msg.equals(MULTIPLY)) {
			getContext().unbecome();
			getContext().become(multiply);
		} else if (msg.equals(DIVIDE)) {
			getContext().unbecome();
			getContext().become(divide);
		}
	}
}

File: ActorMessagingExample/src/main/scala/org/akka/essentials/hotswap/CalculatorActor.scala
package org.akka.essentials.hotswap
import akka.actor.Actor
import akka.actor.ActorLogging

case class ADD
case class SUBTRACT
case class MULTIPLY
case class DIVIDE
case class Numbers(first: Integer, second: Integer)

class CalculatorActor extends Actor with ActorLogging {

	override def preStart() {
		context.become(add)
	}

	def receive = {
		case _ => log.info("unknown message")
	}

	def add: Receive = {
		case number: Numbers => sender ! number.first + number.second
		case "Minus" => 
		case msg: Object => checkAndUpdate(msg.toString())
	}

	def subtract: Receive = {
		case number: Numbers => sender ! number.first - number.second
		case msg: Object => checkAndUpdate(msg.toString())
	}

	def multiply: Receive = {
		case number: Numbers => sender ! number.first * number.second
		case msg: Object => checkAndUpdate(msg.toString())
	}

	def divide: Receive = {
		case number: Numbers => sender ! number.first / number.second
		case msg: Object => checkAndUpdate(msg.toString())
	}

	def checkAndUpdate: Receive = {
		case "ADD" =>
			context.unbecome()
			context.become(add)
		case "SUBTRACT" =>
			context.unbecome()
			context.become(subtract)
		case "MULTIPLY" =>
			context.unbecome()
			context.become(multiply)
		case "DIVIDE" =>
			context.unbecome()
			context.become(divide)
	}
}